qiskit_experiments/library/tomography/basis/local_basis.py

```diff
@@ -0,0 +1,413 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Circuit basis for tomography preparation and measurement circuits
+"""
+import functools
+from typing import Sequence, Optional, Tuple, Union, List, Dict
+import numpy as np
+from qiskit.circuit import QuantumCircuit, Instruction
+from qiskit.quantum_info import DensityMatrix, Statevector, Operator, SuperOp
+from qiskit.quantum_info.operators.channel.quantum_channel import QuantumChannel
+from qiskit.exceptions import QiskitError
+from .base_basis import PreparationBasis, MeasurementBasis
+
+
+class LocalPreparationBasis(PreparationBasis):
+    """Local tensor-product preparation basis.
+
+    This basis consists of a set of 1-qubit instructions which
+    are used to define a tensor-product basis on N-qubits.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        instructions: Optional[Sequence[Instruction]] = None,
+        default_states: Optional[Sequence[Union[Statevector, DensityMatrix]]] = None,
+        qubit_states: Optional[Dict[int, Sequence[Union[Statevector, DensityMatrix]]]] = None,
+    ):
+        """Initialize a fitter preparation basis.
+
+        Args:
+            name: a name to identity the basis.
+            instructions: list of 1-qubit instructions for preparing states
+                          from the :math:`|0^{\\otimes n}\\rangle` state.
+            default_states: Optional, default density matrices prepared by the
+                            input instructions. If None these will be determined by
+                            ideal simulation of the preparation circuits. These
+                            values will be used for any qubit args not explicitly
+                            defined in the qubit_states.
+            qubit_states: Optional, density matrices prepared by the input
+                          instructions on specified qubits. If None the default
+                          states will be used for all qubits.
+
+        Raises:
+            QiskitError: if input states or instructions are not valid.
+        """
+        if instructions is None and default_states is None and qubit_states is None:
+            raise QiskitError(
+                "LocalPreparationBasis must define at least one of instructions, "
+                "default_states, or qubit_states."
+            )
+        super().__init__(name)
+
+        # Internal variables
+        self._instructions = tuple()
+        self._size = None
+        self._default_states = None
+        self._default_dim = None
+        self._qubit_states = {}
+        self._qubit_dim = {}
+        self._qubits = set()
+        self._custom_defaults = True
+
```

Comment:
Something like that would have to go into the base classes to work for every basis, so probably another abstract method like `label(index, qubits) -> str`. I'll have a think about it, but it could be a follow up since isn't necessary for any part of tomography experiment.